File: src/sample/lua/data/sample.lua
function name(type)
    local constness = tonumber(type.constness) == 0 and 'const ' or ''
    local access = tonumber(type.access) == 0 and 'const ' or ''
    local ptr
    if type.indirection == 0 then
        return(tostring("%s%s"):format(access, type.metaclass.name))
    else
        if type.indirection == 1 then
            ptr = "raw"
        elseif type.indirection == 2 then
            ptr = "weak"
        elseif type.indirection == 3 then
            ptr = "ref"
        else
            ptr = "???"
        end
        return(tostring("%s%s<%s%s>"):format(constness,
                                   ptr,
                                   access,
                                   type.metaclass.name))
    end
end

function help(klass)
    function print_method(method)
        for i=1, method.overloads:size(), 1 do
            local overload = method.overloads[i]
            local param_list = {}
            for j=1, overload.params:size(), 1 do
                local param = overload.params[j]
                param_list[#param_list+1] = tostring('%s %s'):format(name(param.type), param.name)
            end
            print(tostring('  %s %s (%s)'):format(name(overload.returnType),
                                                  method.name,
                                                  table.concat(param_list, ', ')))
        end
    end

    print(tostring('class %s'):format(klass.name))
    if klass.constructor then
        print('List of constructors:')
        print_method(klass.constructor)
    end
    print('List of methods:')
    for i=1, klass.methods:size(), 1 do
        local method = klass.methods[i]
        print_method(method)
    end
    print('List of properties:')
    for i=1, klass.properties:size(), 1 do
        local property = klass.properties[i]
        print(tostring(' %s %s'):format(name(property.type), property.name))
    end
    print('List of objects:')
    local object = klass.objects
    while object do
        print(tostring(' %s'):format(object.name))
        object = object.next
    end
end

p = plugin('sample.lua')

help(Motor.Meta.Class.ClassType.metaclass)
help(p.LuaTest)

local property = p.LuaTest.properties[1]
help(property.type.metaclass)

tuple = property.type.metaclass(1, 2, 3, 4)
print(tuple)
function print_tuple(tuple)
    print(tuple.__index)
    result = '( '
    for i=1, tuple.__type.metaclass.properties:size(), 1 do
        local property = tuple.__type.metaclass.properties[i]
        result = result .. tostring('%s[%s]:%s '):format(name(property.type), property.name, property:get(tuple))
    end
    result = result .. ')'
    print(result)
end

print_tuple(tuple)
tuple._0 = 5
tuple._2 = 7
print_tuple(tuple)

File: src/sample/python/data/sample.py
from py_motor import *

p = Plugin("sample.python")
print(p.Motor.TestCases.Enum.Value1)
print(p.Motor.TestCases.Enum.Value2)
print(p.Motor.TestCases.Enum.Value3)


def motor_name(motor_type: Value) -> str:
    constness = int(motor_type.constness) == 0 and 'const ' or ''
    access = int(motor_type.access) == 0 and 'const ' or ''
    if motor_type.indirection == Motor.Meta.Type.Indirection.Value:
        return '%s%s' % (access, motor_type.metaclass.name)
    else:
        if motor_type.indirection == Motor.Meta.Type.Indirection.RawPtr:
            ptr = 'raw'
        elif motor_type.indirection == Motor.Meta.Type.Indirection.WeakPtr:
            ptr = 'weak'
        elif motor_type.indirection == Motor.Meta.Type.Indirection.RefPtr:
            ptr = 'ref'
        else:
            ptr = '???'
        return '%s%s<%s%s>' % (constness, ptr, access, motor_type.metaclass.name)


def print_method(motor_method: Value) -> None:
    for overload in motor_method.overloads:
        param_list = []
        for param in overload.params:
            param_list.append((motor_name(param.type), param.name))
        print(
            '  %s %s (%s)' % (
                motor_name(overload.returnType), motor_method.name,
                ', '.join(('%s %s' % param for param in param_list)))
        )


def motor_help(klass: Value) -> None:
    print('class %s' % klass.name)
    if klass.constructor:
        print('List of constructors:')
        print_method(klass.constructor)
    print('List of methods:')
    for motor_method in klass.methods:
        print_method(motor_method)
    print('List of properties:')
    for motor_property in klass.properties:
        print(' ', motor_name(motor_property.type), motor_property.name)
    print('List of objects:')
    motor_object = klass.objects
    while motor_object:
        print(' ', motor_object.name)
        motor_object = motor_object.next


if __name__ == '__main__':
    help(Motor.Meta.Class.ClassType.metaclass)
    help(Motor.Meta.Class)
    help(Motor.text)
    help(Motor.DiskFolder)

    sample = Plugin('sample.python')
    c = sample.TestCases.Class(y1=1, x1=3)
    assert c is not None
    c.doStuff(1, 2, True)
    print('%d - %d' % (c.x1, c.y1))
